File: js/src/timing.ts
import { Phase } from './kinds.js';

// Always collected, printed only under CAPTCHA_TIMINGS=1: a budget you have to opt into is one nobody has when the slow solve happens.
export const PRODUCTIVE: ReadonlySet<Phase> = new Set<Phase>([Phase.INFERENCE, Phase.MOUSE]);

export function timingsEnabled(): boolean {
  return process.env.CAPTCHA_TIMINGS === '1';
}

/** Phases attribute, they do not partition: a nested phase counts under both names, and re-entering an open one counts once. */
export class PhaseBudget {
  readonly totals = new Map<Phase, number>();
  readonly counts = new Map<Phase, number>();
  private readonly open: Phase[] = [];
  private readonly t0 = Date.now();

  async phase<T>(name: Phase, fn: () => Promise<T>): Promise<T> {
    if (this.open.includes(name)) return fn();
    this.open.push(name);
    const t0 = Date.now();
    try {
      return await fn();
    } finally {
      this.open.splice(this.open.indexOf(name), 1);
      this.add(name, Date.now() - t0);
    }
  }

  add(name: Phase, ms: number): void {
    this.totals.set(name, (this.totals.get(name) ?? 0) + ms);
    this.counts.set(name, (this.counts.get(name) ?? 0) + 1);
  }

  elapsedMs(): number {
    return Date.now() - this.t0;
  }

  toObject(): Record<string, number> {
    const out: Record<string, number> = {};
    for (const [k, v] of this.totals) out[k] = v;
    out.total = this.elapsedMs();
    return out;
  }

  report(): string {
    const total = this.elapsedMs();
    let useful = 0;
    for (const [k, v] of this.totals) if (PRODUCTIVE.has(k)) useful += v;
    const rows = [...this.totals.entries()].sort((a, b) => b[1] - a[1]);
    const lines = [
      `[BUDGET] solve ${(total / 1000).toFixed(1)}s — ` +
        `${(useful / 1000).toFixed(1)}s useful (${total ? Math.round((100 * useful) / total) : 0}%), ` +
        `${((total - useful) / 1000).toFixed(1)}s waiting`,
    ];
    for (const [name, ms] of rows) {
      const tag = PRODUCTIVE.has(name) ? '*' : ' ';
      lines.push(
        `[BUDGET] ${tag} ${name.padEnd(22)} ${(ms / 1000).toFixed(2).padStart(6)}s  x${this.counts.get(name)}`,
      );
    }
    let attributed = 0;
    for (const v of this.totals.values()) attributed += v;
    if (total - attributed > 50) {
      lines.push(`[BUDGET]   ${'(unattributed)'.padEnd(22)} ${((total - attributed) / 1000).toFixed(2).padStart(6)}s`);
    }
    return lines.join('\n');
  }
}

File: python/src/captchakraken/timing.py
"""Where one solve's wall-clock went, by phase. Always collected, printed under CAPTCHA_TIMINGS=1.

Always on because a budget you have to opt into is one nobody has when the slow solve happens.
"""

from __future__ import annotations

import os
import time
from contextlib import contextmanager
from typing import Dict, Iterator, List

from .kinds import Phase

# Only a second the model is thinking in or the mouse is travelling in is worth spending; every other
# phase is the driver waiting on a clock, and a candidate for deletion or overlap.
PRODUCTIVE = frozenset({Phase.INFERENCE, Phase.MOUSE})


def timings_enabled() -> bool:
    return os.getenv("CAPTCHA_TIMINGS", "0") == "1"


class PhaseBudget:
    def __init__(self) -> None:
        self.totals: Dict[Phase, float] = {}
        self.counts: Dict[Phase, int] = {}
        self._open: List[Phase] = []
        self._t0 = time.perf_counter()

    @contextmanager
    def phase(self, name: Phase) -> Iterator[None]:
        # Only the outermost entry of a name accumulates; nesting under another name counts under both.
        if name in self._open:
            yield
            return
        self._open.append(name)
        t0 = time.perf_counter()
        try:
            yield
        finally:
            self._open.remove(name)
            self.add(name, (time.perf_counter() - t0) * 1000.0)

    def add(self, name: Phase, ms: float) -> None:
        self.totals[name] = self.totals.get(name, 0.0) + ms
        self.counts[name] = self.counts.get(name, 0) + 1

    def elapsed_ms(self) -> float:
        return (time.perf_counter() - self._t0) * 1000.0

    def report(self) -> str:
        total = self.elapsed_ms()
        useful = sum(v for k, v in self.totals.items() if k in PRODUCTIVE)
        lines = [f"[BUDGET] solve {total / 1000:.1f}s — {useful / 1000:.1f}s useful "
                 f"({100 * useful / total if total else 0:.0f}%), {(total - useful) / 1000:.1f}s waiting"]
        for name, ms in sorted(self.totals.items(), key=lambda kv: -kv[1]):
            tag = "*" if name in PRODUCTIVE else " "
            lines.append(f"[BUDGET] {tag} {name:22s} {ms / 1000:6.2f}s  x{self.counts[name]}")
        unattributed = total - sum(self.totals.values())
        if unattributed > 50:
            lines.append(f"[BUDGET]   {'(unattributed)':22s} {unattributed / 1000:6.2f}s")
        return "\n".join(lines)
